src/twisted/test/test_defer.py

```diff
@@ -1678,6 +1681,42 @@ def test_fromCoroutineRequiresCoroutine(self) -> None:
         for thing in thingsThatAreNotCoroutines:
             self.assertRaises(defer.NotACoroutineError, Deferred.fromCoroutine, thing)
 
+    @pyunit.skipIf(_PYPY, "GC works differently on PyPy.")
+    def test_canceller_circular_reference(self) -> None:
+        """
+        Test that a circular reference between a `Deferred` and its canceller
+        is broken when the deferred is resolved.
+        """
+
+        # Create a canceller and weak reference to track if its been freed.
+        canceller = _TestCircularCanceller()
+        weakCanceller = weakref.ref(canceller)
+
+        # Create a Deferred with the above canceller, adding a reference to the
+        # deferred to the canceller to create the circular reference.
+        deferred: Deferred[Any] = Deferred(canceller)
+        canceller.deferred = deferred
+
+        deferred.callback(None)
```

Comment:
Maybe it is worth extending the test to check that the canceller is "free" even when the deferred doesn't have a "final" result yet.

```suggestion
        pendingResult = Deferred()
        deferred.callback(pendingResult)
```